rust/protocol/src/address.rs

```diff
@@ -1,26 +1,57 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2021 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+//! Where to send a [sealed_sender] message to.
+
+// These are used in docstrings.
+#[allow(unused_imports)]
+use crate::sealed_sender::{self, UnidentifiedSenderMessageContent};
+
 use std::fmt;
 
+/// The type used in memory to represent a *device*, i.e. a particular Signal client instance which
+/// represents some user.
+///
+/// Used in [ProtocolAddress].
+pub type DeviceId = u32;
+
+/// The target of an [UnidentifiedSenderMessageContent].
 #[derive(Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
 pub struct ProtocolAddress {
     name: String,
-    device_id: u32,
+    device_id: DeviceId,
 }
 
 impl ProtocolAddress {
-    pub fn new(name: String, device_id: u32) -> Self {
+    /// Create a new address.
+    ///
+    /// - `name` defines a user's public identity, and therefore must be globally unique to that
+    /// user.
+    /// - Each Signal client instance then has its own `device_id`, which must be unique among
+    ///   all clients for that user.
+    pub fn new(name: String, device_id: DeviceId) -> Self {
         ProtocolAddress { name, device_id }
     }
 
+    /// A unique identifier for the target user.
+    ///
+    /// In the Signal app, this is currently a phone number.
+    #[inline]
     pub fn name(&self) -> &str {
         &self.name
     }
 
-    pub fn device_id(&self) -> u32 {
+    /// An identifier representing a particular Signal client instance to send to.
+    ///
+    /// For example, if a user has set up Signal on both their phone and laptop, any
+    /// [UnidentifiedSenderMessageContent] sent to the user will only go to a single
+    /// device. Handling this is part of the [Sesame] algorithm.
+    ///
+    /// [Sesame]: https://signal.org/docs/specifications/sesame/#assumptions
```

Comment:
I misread this at first as "if a user sends a message on Signal, it goes to exactly one of the recipient's devices", as opposed to "a single Signal message goes to a particular device, so when a user sends a message they're actually sending a message to *every* device", which I think means the example is potentially more confusing than helpful. It's not really Sesame-related either—that's everything to do with "pre-keys" rather than multiple devices per user.

Reply:
Thank you for this thorough feedback! I will try to correct this.